File: lib/iriq/inflector.rb
require "set"

module Iriq
  # Singularization with a swappable adapter.
  #
  # By default uses ActiveSupport's inflector if it can be required, otherwise
  # falls back to BuiltinAdapter. Override globally:
  #
  #   Iriq::Inflector.adapter = MyAdapter   # must respond to .singularize(String)
  #
  # And reset to default with `Iriq::Inflector.reset_adapter!`.
  module Inflector
    # Vocabulary is bounded in practice; cache + cap matches the
    # SegmentClassifier strategy.
    CACHE_MAX = 10_000

    class << self
      def singularize(word)
        cache = (@cache ||= {})
        cached = cache[word]
        return cached if cached

        cache.clear if cache.size >= CACHE_MAX
        cache[word] = adapter.singularize(word)
      end

      def adapter
        @adapter ||= default_adapter
      end

      def adapter=(value)
        @adapter = value
        @cache = {} # different adapter could singularize differently
      end

      def reset_adapter!
        @adapter = nil
        @cache = {}
      end

      def default_adapter
        require "active_support/inflector"
        ActiveSupportAdapter
      rescue LoadError
        BuiltinAdapter
      end
    end

    module ActiveSupportAdapter
      def self.singularize(word)
        ::ActiveSupport::Inflector.singularize(word.to_s)
      end
    end

    # Rule-based English singularizer. Rules are ordered most-specific-first
    # and adapted from ActiveSupport's default inflections.
    module BuiltinAdapter
      IRREGULARS = {
        "people"   => "person",
        "children" => "child",
        "men"      => "man",
        "women"    => "woman",
        "mice"     => "mouse",
        "geese"    => "goose",
        "oxen"     => "ox",
        "feet"     => "foot",
        "teeth"    => "tooth",
        "lives"    => "life",
        "wives"    => "wife",
        "moves"    => "move",
        "zombies"  => "zombie",
        # latin/greek plurals that don't fit a clean suffix rule
        "indices"  => "index",
        "vertices" => "vertex",
        # -f/-fe words where the stem doesn't end in l/r/i
        "leaves"   => "leaf",
        "calves"   => "calf",
        "halves"   => "half",
        "loaves"   => "loaf",
        "hooves"   => "hoof",
      }.freeze

      UNCOUNTABLE = Set.new(%w[
        news fish sheep deer series species equipment information
        money rice jeans police data media
      ]).freeze

      # [pattern, replacement] — first match wins.
      RULES = [
        [/(quiz)zes$/i,                       '\1'],
        [/(matri|appendi)ces$/i,              '\1x'],
        [/(ox)en$/i,                          '\1'],
        [/(alias|status)(es)?$/i,             '\1'],
        [/(octop|vir)(us|i)$/i,               '\1us'],
        [/(cris|ax|test)es$/i,                '\1is'],
        [/(shoe)s$/i,                         '\1'],
        [/(bus)(es)?$/i,                      '\1'],
        [/([ml])ice$/i,                       '\1ouse'],
        [/(x|ch|ss|sh)es$/i,                  '\1'],
        [/(m)ovies$/i,                        '\1ovie'],
        [/(s)eries$/i,                        '\1eries'],
        [/([^aeiouy]|qu)ies$/i,               '\1y'],
        [/([lr])ves$/i,                       '\1f'],
        [/(tive)s$/i,                         '\1'],
        [/(hive)s$/i,                         '\1'],
        [/([^f])ves$/i,                       '\1fe'],
        [/((a)naly|(b)a|(d)iagno|(p)arenthe|(p)rogno|(s)ynop|(t)he)ses$/i, '\1sis'],
        [/([ti])a$/i,                         '\1um'],
        [/(n)ews$/i,                          '\1ews'],
        [/(o)es$/i,                           '\1'],
        [/(ss)$/i,                            '\1'],
        [/s$/i,                               ''],
      ].freeze

      def self.singularize(word)
        return word if word.nil? || word.empty?

        lower = word.downcase
        return word if UNCOUNTABLE.include?(lower)

        if (irr = IRREGULARS[lower])
          return preserve_case(word, irr)
        end

        RULES.each do |pattern, replacement|
          if word.match?(pattern)
            return word.sub(pattern, replacement)
          end
        end

        word
      end

      def self.preserve_case(original, lowered)
        if original == original.upcase
          lowered.upcase
        elsif original[0] == original[0].upcase
          lowered.sub(/\A./, &:upcase)
        else
          lowered
        end
      end
    end
  end
end

File: rust/iriq/src/inflector.rs
// Built-in English singularizer. Rules adapted from ActiveSupport defaults
// via the Go port. Mirrors the same precedence and irregular/uncountable
// lists so output is bit-identical across runtimes.

use once_cell::sync::Lazy;
use regex::Regex;
use std::collections::{HashMap, HashSet};
use std::sync::Mutex;

const CACHE_MAX: usize = 10_000;

static IRREGULARS: Lazy<HashMap<&'static str, &'static str>> = Lazy::new(|| {
    let mut m = HashMap::new();
    m.insert("people", "person");
    m.insert("children", "child");
    m.insert("men", "man");
    m.insert("women", "woman");
    m.insert("mice", "mouse");
    m.insert("geese", "goose");
    m.insert("oxen", "ox");
    m.insert("feet", "foot");
    m.insert("teeth", "tooth");
    m.insert("lives", "life");
    m.insert("wives", "wife");
    m.insert("moves", "move");
    m.insert("zombies", "zombie");
    m.insert("indices", "index");
    m.insert("vertices", "vertex");
    m.insert("leaves", "leaf");
    m.insert("calves", "calf");
    m.insert("halves", "half");
    m.insert("loaves", "loaf");
    m.insert("hooves", "hoof");
    m
});

static UNCOUNTABLE: Lazy<HashSet<&'static str>> = Lazy::new(|| {
    [
        "news", "fish", "sheep", "deer", "series", "species",
        "equipment", "information", "money", "rice", "jeans",
        "police", "data", "media",
    ]
    .iter()
    .copied()
    .collect()
});

struct Rule {
    re: Regex,
    repl: &'static str,
}

static RULES: Lazy<Vec<Rule>> = Lazy::new(|| {
    vec![
        Rule { re: Regex::new(r"(?i)(quiz)zes$").unwrap(), repl: "${1}" },
        Rule { re: Regex::new(r"(?i)(matri|appendi)ces$").unwrap(), repl: "${1}x" },
        Rule { re: Regex::new(r"(?i)(ox)en$").unwrap(), repl: "${1}" },
        Rule { re: Regex::new(r"(?i)(alias|status)(es)?$").unwrap(), repl: "${1}" },
        Rule { re: Regex::new(r"(?i)(octop|vir)(us|i)$").unwrap(), repl: "${1}us" },
        Rule { re: Regex::new(r"(?i)(cris|ax|test)es$").unwrap(), repl: "${1}is" },
        Rule { re: Regex::new(r"(?i)(shoe)s$").unwrap(), repl: "${1}" },
        Rule { re: Regex::new(r"(?i)(bus)(es)?$").unwrap(), repl: "${1}" },
        Rule { re: Regex::new(r"(?i)([ml])ice$").unwrap(), repl: "${1}ouse" },
        Rule { re: Regex::new(r"(?i)(x|ch|ss|sh)es$").unwrap(), repl: "${1}" },
        Rule { re: Regex::new(r"(?i)(m)ovies$").unwrap(), repl: "${1}ovie" },
        Rule { re: Regex::new(r"(?i)(s)eries$").unwrap(), repl: "${1}eries" },
        Rule { re: Regex::new(r"(?i)([^aeiouy]|qu)ies$").unwrap(), repl: "${1}y" },
        Rule { re: Regex::new(r"(?i)([lr])ves$").unwrap(), repl: "${1}f" },
        Rule { re: Regex::new(r"(?i)(tive)s$").unwrap(), repl: "${1}" },
        Rule { re: Regex::new(r"(?i)(hive)s$").unwrap(), repl: "${1}" },
        Rule { re: Regex::new(r"(?i)([^f])ves$").unwrap(), repl: "${1}fe" },
        Rule {
            re: Regex::new(
                r"(?i)((a)naly|(b)a|(d)iagno|(p)arenthe|(p)rogno|(s)ynop|(t)he)ses$",
            )
            .unwrap(),
            repl: "${1}sis",
        },
        Rule { re: Regex::new(r"(?i)([ti])a$").unwrap(), repl: "${1}um" },
        Rule { re: Regex::new(r"(?i)(n)ews$").unwrap(), repl: "${1}ews" },
        Rule { re: Regex::new(r"(?i)(o)es$").unwrap(), repl: "${1}" },
        Rule { re: Regex::new(r"(?i)(ss)$").unwrap(), repl: "${1}" },
        Rule { re: Regex::new(r"(?i)s$").unwrap(), repl: "" },
    ]
});

static CACHE: Lazy<Mutex<HashMap<String, String>>> = Lazy::new(|| Mutex::new(HashMap::new()));

pub fn singularize(word: &str) -> String {
    {
        let mut cache = CACHE.lock().unwrap();
        if let Some(v) = cache.get(word) {
            return v.clone();
        }
        if cache.len() >= CACHE_MAX {
            cache.clear();
        }
    }
    let out = singularize_inner(word);
    let mut cache = CACHE.lock().unwrap();
    cache.insert(word.to_string(), out.clone());
    out
}

fn singularize_inner(word: &str) -> String {
    if word.is_empty() {
        return word.to_string();
    }
    let lower = word.to_lowercase();
    if UNCOUNTABLE.contains(lower.as_str()) {
        return word.to_string();
    }
    if let Some(&irr) = IRREGULARS.get(lower.as_str()) {
        return preserve_case(word, irr);
    }
    for r in RULES.iter() {
        if r.re.is_match(word) {
            return r.re.replace(word, r.repl).into_owned();
        }
    }
    word.to_string()
}

fn preserve_case(original: &str, lowered: &str) -> String {
    if original == original.to_uppercase() {
        return lowered.to_uppercase();
    }
    if let Some(first) = original.chars().next() {
        if first.is_uppercase() {
            let mut chars = lowered.chars();
            if let Some(c) = chars.next() {
                let upper: String = c.to_uppercase().collect();
                let rest: String = chars.collect();
                return upper + &rest;
            }
        }
    }
    lowered.to_string()
}
